CorsixTH/Lua/entities/humanoids/patient.lua

```diff
@@ -1122,39 +1122,36 @@ end
 !param meander_before_resume (bool) Indicator to insert an additional action before resuming.
 ]]
 function Patient:interruptAndRequeueAction(current_action, queue_pos, meander_before_resume)
-  if current_action.name == "idle" or current_action.name == "walk" then
-    local requeue_action = {
-        name = current_action.name,
-        x = current_action.x,
-        y = current_action.y,
-        must_happen = current_action.must_happen,
-        is_entering = current_action.is_entering,
-    }
-    self:queueAction(requeue_action, queue_pos)
-    -- If we were idling, also go away a little before continuing with
-    -- that important action.
-    if current_action.name == "idle" and meander_before_resume then
+  -- Rebuild the current action.
+  local requeue_action
+  if current_action.name == "idle" then
+    requeue_action = IdleAction():setMustHappen(current_action.must_happen)
+    requeue_action:setDirection(current_action.direction)
+
+    -- Go away a little before continuing with that important action.
+    if meander_before_resume then
       self:queueAction(MeanderAction():setCount(1), queue_pos)
-    -- if walking, and action was already started, use the original must_happen
-    elseif current_action.name == "walk" then
-      if current_action.saved_must_happen ~= nil then
-        requeue_action.must_happen = current_action.saved_must_happen
-      end
-      -- need to copy the reserve_on_resume, otherwise the new queued action will not
-      -- unreserve on interrupt
-      requeue_action.reserve_on_resume = current_action.reserve_on_resume
     end
+  elseif current_action.name == "walk" then
+    requeue_action = WalkAction(current_action.x, current_action.y)
+    requeue_action:setMustHappen(current_action.must_happen)
+
+    -- need to copy the reserve_on_resume, otherwise the new queued action will not
+    -- unreserve on interrupt
+    requeue_action.reserve_on_resume = current_action.reserve_on_resume
```

Comment:
Did you mean to leave this one?

Reply:
As in not making a `WalkAction:setReserveOnResume`?   Yes, at least for now.


The action code has callbacks, interrupts, requeueing and a zillion flags like this one for customizing and steering behavior all over the place. To me, it says that it had the nice but naive idea of nice sunny weather sequences of animations, and then it turned out to be much less sunny causing a lot of the complexity.
I don't know all the cases and handling and work-arounds yet, since the structure is hard to read with all this
callback glue. Even if I know it, there is currently no common data storage in place, as functions don't share data except by parameter.

So for now, I just try to move code to a more OO-like solution starting by sticking things into or at least near the `HumanoidAction` derived classes, where #2087 is an easy case. I don't know if that works for everything, nor do I really know where it will lead to, but I see no other way to get forward.